core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2254,7 +2256,21 @@ class Log(@volatile var dir: File,
    */
   private[log] def replaceSegments(newSegments: Seq[LogSegment], oldSegments: Seq[LogSegment], isRecoveredSwapFile: Boolean = false): Unit = {
     lock synchronized {
+      if (newSegments.size > 1 && newSegments.exists(_.size == 0))
+        throw new IllegalArgumentException(s"Expected only one cleaned segment should be created after cleaning, but we got $newSegments")
       val sortedNewSegments = newSegments.sortBy(_.baseOffset)
+
+      val (emptyCleanedSegments, nonEmptyCleanedSegments) = sortedNewSegments.partition(_.size == 0)
+
+      if (emptyCleanedSegments.nonEmpty) {
+        val baseOffset = segments.firstKey()
+        if (emptyCleanedSegments.head.baseOffset == baseOffset) {
+          val newLogStartOffset = Math.max(logStartOffset, segments.higherKey(baseOffset))
```

Comment:
Hmm, this doesn't seem quite right. There could be multiple segments in oldSegments that the new segment is trying to replace. In that case, we want to get the higher key of the last old segment.

Reply:
@junrao Sorry for the long delay. Will take care of the comments ASAP.